### PR TITLE
[DRAFT] Add write endpoints for all US Core profiles

### PR DESCRIPTION
**Pull request by Robert-W**
**What's here**
* Adds two write endpoints to all US Core profiles
  * `/:version/[profile]` as a `post` route for creates
  * `/:version/[profile]/:id` as a `put` route for updates and creates
* Added responseUtils to clean up all the controller classes and reduce the duplicated code
* Added additional write specific arguments
* Updated all controllers and services (with mock services)

**Tasks Still Remaining**
* Update Documentation (needs to happen in the Wiki after merging)
* Add in additional security after talking to Dan
* Add more unit tests

**Additional PR's**
* We should start on another PR soon to add support for deletes but this will require a little more research to make sure we understand the required scopes and how to handle refs and other things.

### CODE REVIEW

File: src/server/profiles/allergyintolerance/allergyintolerance.config.js [hunk 1]
```diff
@@ -14,6 +14,17 @@ const scopes = [
 	'allergyintolerance/*.read'
 ];
 
+const write_scopes = [
```
Comment (jonterrylee) [inline src/server/profiles/allergyintolerance/allergyintolerance.config.js]: guess theres no really pretty way to separate this out...

File: src/server/profiles/allergyintolerance/allergyintolerance.controller.js [hunk 2]
```diff
@@ -1,45 +1,23 @@
...
-			})
-			.catch((err) => {
-				next(errors.internal(err.message, version));
-			});
+			.then((results) =>
+				responseUtils.handleBundleReadResponse( res, version, AllergyIntolerance, results, {
+					resourceUrl: config.auth.resourceServer
+				})
+			)
+			.catch((err) => next(errors.internal(err.message, version)));
```
Comment (jonterrylee) [inline src/server/profiles/allergyintolerance/allergyintolerance.controller.js]: are we logging the errors anywhere?  service?  If not can we log it here so we don't lose the stack trace

File: src/server/profiles/allergyintolerance/allergyintolerance.config.js (hunk 1 above)
Comment (Robert-W) [reply src/server/profiles/allergyintolerance/allergyintolerance.config.js]: yea not really, we can pull them out into their own separate module but then have to pass strings in for each module, so not sure it is much better, may save a couple of lines though.

File: src/server/profiles/allergyintolerance/allergyintolerance.controller.js (hunk 2 above)
Comment (Robert-W) [reply src/server/profiles/allergyintolerance/allergyintolerance.controller.js]: Ok added

File: src/server/profiles/allergyintolerance/allergyintolerance.config.js (hunk 1 above)
Comment (Robert-W) [reply src/server/profiles/allergyintolerance/allergyintolerance.config.js]: How would you feel about something like this:

```javascript
// In a common.scopes.js file
module.exports.read_scopes = profile => {
	let profile_lower = profile.toLowerCase();

	return [
		'user/*.*',
		`user/${profile_lower}.*`,
		`user/${profile_lower}.read`,
		'user/*.read',
		`${profile_lower}/*.*`,
		`${profile_lower}/${profile}.*`,
		`${profile_lower}/${profile}.read`,
		`${profile_lower}/*.read`
	];
};

// In allergyintolerance.config.js
const read_scopes = common_scopes.read_scopes('AllergyIntolerance');
```

**Pull request merged**